[GEN-1164] add workflow to build docker image  and push it to ghcr

danlu1 opened the pull request:
Problem:

There are limited admin seats that we can have on dockerHub.

Solution:
Switch to pushing our docker images to ghcr

Output:

The docker image and cached image can be viewed [here](https://github.com/Sage-Bionetworks/Genie/pkgs/container/genie/versions).

danlu1 committed:
add workflow to build aocker image  and push it to ghcr

thomasyu888 commented on .github/workflows/docker_build.yml:
@danlu1 this is great, but can we add this to our existing github action?  We don't necessarily want to build if the tests fail

danlu1 replied:
Okay. I will work on this.

rxu17 commented on .github/workflows/docker_build.yml:
```diff
@@ -0,0 +1,53 @@
+        id: meta
+        uses: docker/metadata-action@v5
+        with:
+          images: ${{ env.REGISTRY }}/${{ env.IMAGE_NAME }}
```
Could you add the following under `images:`
```
tags: |
    type=semver,pattern={{version}}
    type=ref,event=branch
    ref=main
    ref=develop
    type=sha
    latest
```
This will make sure we have docker image for the following tags `develop`, `main`, `latest`, and version of the tag I don't know if we need to have docker images created and tagged per feature branch as well.

danlu1 replied:
@rxu17 Thx for the comments. Could you plz point me to the doc describing using `ref=` in the tags sections? My original thoughts on achieving this goal are adding `main` branch to the push event and only use     
```
type=semver,pattern={{version}}
type=ref,event=branch
type=ref,event=tag
type=raw,value=latest,enable={{is_default_branch}} # to conditionally tag with latest for the default branch
``` 
for tags section.

danlu1 replied:
Based on our discussion on slack, we would like to add docker build step in `ci` workflow, create a docker image for develop, tag, and feature branch, and to use Registry cache since gha has a 10GB per repo storage limit.

danlu1 committed:
add build container step to ci workflow

danlu1 committed:
uncomment needs for build-container

danlu1 committed:
remove docker_build.yml

danlu1 committed:
rename build-docker step

danlu1 committed:
Update ci.yml

Remove tags section in docker/metadata-action since only push event will generate tag for branch. But we want branch based cache.

danlu1 committed:
Update ci.yml

Disable provenance to avoid caching manifest

danlu1 committed:
build docker image when it is not a  pull request

danlu1 committed:
make push event on develop and feature branches

rxu17 commented on .github/workflows/ci.yml:
```diff
@@ -5,15 +5,18 @@ name: build
   push:
-    branches:
-      - develop
+    branches: [develop, 'GEN*', 'gen*']
```
Just a comment to note that `test` and `lint` will run for every push now to the feature branches regardless of whether there is a PR or not. Nothing to fix here.

danlu1 replied:
Yes, right now, push to develop, branches with gen/GEN prefix, all PRs and release will trigger the test and lint job.

rxu17 commented on .github/workflows/ci.yml:
```diff
@@ -5,15 +5,18 @@ name: build
   push:
-    branches:
-      - develop
+    branches: [develop, 'GEN*', 'gen*']
```
Actually could you also add `main` to the branches here? I don't know if your current setup for ghcr will always create a docker image that is based on the latest version of the repo like we have for dockerhub? Will this setup create a `latest` tag by default?

rxu17 reviewed (commented):
Thanks for tackling this! Just a question

danlu1 committed:
Update ci.yml

Add main branch to push event branches.

rxu17 reviewed (approved):
LGTM!

thomasyu888 reviewed (commented):
🔥 thanks for the great work!  Great review
